// @creditkarma/thrift-parser/dist/main/keywords.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("./types");
exports.KEYWORDS = {
    namespace: types_1.SyntaxType.NamespaceKeyword,
    include: types_1.SyntaxType.IncludeKeyword,
    cpp_include: types_1.SyntaxType.CppIncludeKeyword,
    const: types_1.SyntaxType.ConstKeyword,
    struct: types_1.SyntaxType.StructKeyword,
    service: types_1.SyntaxType.ServiceKeyword,
    extends: types_1.SyntaxType.ExtendsKeyword,
    throws: types_1.SyntaxType.ThrowsKeyword,
    typedef: types_1.SyntaxType.TypedefKeyword,
    union: types_1.SyntaxType.UnionKeyword,
    enum: types_1.SyntaxType.EnumKeyword,
    senum: types_1.SyntaxType.SenumKeyword,
    list: types_1.SyntaxType.ListKeyword,
    set: types_1.SyntaxType.SetKeyword,
    map: types_1.SyntaxType.MapKeyword,
    double: types_1.SyntaxType.DoubleKeyword,
    i8: types_1.SyntaxType.I8Keyword,
    i16: types_1.SyntaxType.I16Keyword,
    i32: types_1.SyntaxType.I32Keyword,
    i64: types_1.SyntaxType.I64Keyword,
    exception: types_1.SyntaxType.ExceptionKeyword,
    binary: types_1.SyntaxType.BinaryKeyword,
    bool: types_1.SyntaxType.BoolKeyword,
    byte: types_1.SyntaxType.ByteKeyword,
    required: types_1.SyntaxType.RequiredKeyword,
    optional: types_1.SyntaxType.OptionalKeyword,
    string: types_1.SyntaxType.StringKeyword,
    true: types_1.SyntaxType.TrueKeyword,
    false: types_1.SyntaxType.FalseKeyword,
    void: types_1.SyntaxType.VoidKeyword,
    oneway: types_1.SyntaxType.OnewayKeyword,
};
